tests/Avalonia.Controls.UnitTests/Utils/HotKeyManagerTests.cs

```diff
@@ -50,10 +55,119 @@ public void HotKeyManager_Should_Register_And_Unregister_Key_Binding()
+                    tl.Content = null;
+                    tl.ApplyTemplate();
+                })();
 
+
+                // The button should be collected since it's detached from the listbox
+                GC.Collect();
+                GC.WaitForPendingFinalizers();
+                GC.Collect();
+                GC.WaitForPendingFinalizers();
+
+                Assert.Null(reference?.Target);
             }
         }
 
+        [Fact]
+        public void HotKeyManager_Should_Release_Reference_When_Control_In_Item_Template_Detached()
+        {
+            using (UnitTestApplication.Start(TestServices.StyledWindow))
+            {
+                var styler = new Mock<Styler>();
+
+                AvaloniaLocator.CurrentMutable
+                    .Bind<IWindowingPlatform>().ToConstant(new WindowingPlatformMock())
+                    .Bind<IStyler>().ToConstant(styler.Object);
+
+                var gesture1 = new KeyGesture(Key.A, KeyModifiers.Control);
+
+                var weakReferences = new List<WeakReference>();
+                var tl = new Window { SizeToContent = SizeToContent.WidthAndHeight, IsVisible = true };
+                var lm = tl.LayoutManager;
+
+                var keyGestures = new AvaloniaList<KeyGesture> { gesture1 };
+                var listBox = new ListBox
+                {
+                    Width = 100,
+                    Height = 100,
+                    VirtualizationMode = ItemVirtualizationMode.None,
+                    // Create a button with binding to the KeyGesture in the template and add it to references list
+                    ItemTemplate = new FuncDataTemplate(typeof(KeyGesture), (o, scope) =>
+                    {
+                        var keyGesture = o as KeyGesture;
+                        var button = new Button
+                        {
+                            DataContext = keyGesture, [!Button.HotKeyProperty] = new Binding("")
+                        };
+                        weakReferences.Add(new WeakReference(button, true));
+                        return button;
+                    })
+                };
+                // Add the listbox and render it
+                tl.Content = listBox;
+                lm.ExecuteInitialLayoutPass();
+                listBox.Items = keyGestures;
+                lm.ExecuteLayoutPass();
+
+                // Let the button detach when clearing the source items
+                keyGestures.Clear();
+                lm.ExecuteLayoutPass();
+                
+                // Add it again to double check,and render
+                keyGestures.Add(gesture1);
+                lm.ExecuteLayoutPass();
+                
+                keyGestures.Clear();
+                lm.ExecuteLayoutPass();
+                
+                // The button should be collected since it's detached from the listbox
+                GC.Collect();
```

Comment:
just curious, why are we forcing GC to do stuff here?

Reply:
The leak was that any control that used HotKey never got cleaned up by GC even after it detaches and is not referenced by anything. This test just makes sure that the hotkey manager does not keep references of the control after it detaches from logical tree, so I need to make the GC collect it to confirm that nothing reference the control anymore.